File: API/4210_Huge_Income/CS/HugeIncomeStrategy.cs
namespace StockSharp.Samples.Strategies;

using System;
using System.Collections.Generic;
using System.Linq;

using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

/// <summary>
/// Port of the MetaTrader expert advisor "Huge Income".
/// Exploits intraday extensions away from the daily open and closes positions ahead of the configured market close.
/// </summary>
public class HugeIncomeStrategy : Strategy
{
	private readonly StrategyParam<decimal> _tradeVolume;
	private readonly StrategyParam<decimal> _minimumRangePips;
	private readonly StrategyParam<int> _buyCutoffHour;
	private readonly StrategyParam<int> _sellCutoffHour;
	private readonly StrategyParam<int> _marketCloseHour;
	private readonly StrategyParam<DataType> _candleType;

	private DateTime? _currentDay;
	private decimal _dailyOpen;
	private decimal _dailyHigh;
	private decimal _dailyLow;
	private bool _hasDailyData;

	/// <summary>
	/// Initializes a new instance of <see cref="HugeIncomeStrategy"/>.
	/// </summary>
	public HugeIncomeStrategy()
	{
		_tradeVolume = Param(nameof(TradeVolume), 0.1m)
			.SetGreaterThanZero()
			.SetDisplay("Trade Volume", "Order size used for entries", "General");

		_minimumRangePips = Param(nameof(MinimumRangePips), 15m)
			.SetNotNegative()
			.SetDisplay("Minimum Range (pips)", "Minimum daily distance from the open before trading", "Filters");

		_buyCutoffHour = Param(nameof(BuyCutoffHour), 22)
			.SetNotNegative()
			.SetLessOrEqual(23)
			.SetDisplay("Buy Cutoff Hour", "Latest hour when new long trades are allowed", "Timing");

		_sellCutoffHour = Param(nameof(SellCutoffHour), 16)
			.SetNotNegative()
			.SetLessOrEqual(23)
			.SetDisplay("Sell Cutoff Hour", "Latest hour when new short trades are allowed", "Timing");

		_marketCloseHour = Param(nameof(MarketCloseHour), 23)
			.SetNotNegative()
			.SetLessOrEqual(23)
			.SetDisplay("Market Close Hour", "Hour when all open positions must be closed", "Timing");

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(15).TimeFrame())
			.SetDisplay("Candle Type", "Timeframe used to build daily statistics", "General");
	}

	/// <summary>
	/// Order volume used for market entries.
	/// </summary>
	public decimal TradeVolume
	{
		get => _tradeVolume.Value;
		set => _tradeVolume.Value = value;
	}

	/// <summary>
	/// Minimum distance in pips required between the daily open and the daily extreme.
	/// </summary>
	public decimal MinimumRangePips
	{
		get => _minimumRangePips.Value;
		set => _minimumRangePips.Value = value;
	}

	/// <summary>
	/// Latest hour when a new long position can be initiated.
	/// </summary>
	public int BuyCutoffHour
	{
		get => _buyCutoffHour.Value;
		set => _buyCutoffHour.Value = value;
	}

	/// <summary>
	/// Latest hour when a new short position can be initiated.
	/// </summary>
	public int SellCutoffHour
	{
		get => _sellCutoffHour.Value;
		set => _sellCutoffHour.Value = value;
	}

	/// <summary>
	/// Hour at which all open trades are forcefully closed.
	/// </summary>
	public int MarketCloseHour
	{
		get => _marketCloseHour.Value;
		set => _marketCloseHour.Value = value;
	}

	/// <summary>
	/// Candle type that drives intraday updates.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();

		_currentDay = null;
		_dailyOpen = 0m;
		_dailyHigh = 0m;
		_dailyLow = 0m;
		_hasDailyData = false;
	}

	/// <inheritdoc />
	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		var subscription = SubscribeCandles(CandleType);
		subscription
			.Bind(ProcessCandle)
			.Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle)
	{
		// Work only with finished candles to avoid partial-day noise.
		if (candle.State != CandleStates.Finished)
			return;

		UpdateDailyLevels(candle);

		if (!IsFormedAndOnlineAndAllowTrading())
			return;

		if (!_hasDailyData)
			return;

		var candleHour = candle.CloseTime.Hour;

		// Always protect open exposure near the configured market close.
		if (candleHour >= MarketCloseHour && Position != 0m)
		{
			ClosePosition();
			return;
		}

		if (!CanOpenNewTrade())
			return;

		var closePrice = candle.ClosePrice;
		var longRange = _dailyOpen - _dailyLow;
		var shortRange = _dailyHigh - _dailyOpen;
		var requiredRange = GetRequiredRange();

		if (ShouldEnterLong(closePrice, longRange, requiredRange, candleHour))
		{
			BuyMarket(TradeVolume);
			return;
		}

		if (ShouldEnterShort(closePrice, shortRange, requiredRange, candleHour))
		{
			SellMarket(TradeVolume);
		}
	}

	private void UpdateDailyLevels(ICandleMessage candle)
	{
		var candleDate = candle.OpenTime.Date;

		if (_currentDay != candleDate)
		{
			_currentDay = candleDate;
			_dailyOpen = candle.OpenPrice;
			_dailyHigh = candle.HighPrice;
			_dailyLow = candle.LowPrice;
			_hasDailyData = true;
			return;
		}

		if (!_hasDailyData)
			return;

		// Track the running extremes for the active day.
		if (candle.HighPrice > _dailyHigh)
			_dailyHigh = candle.HighPrice;

		if (candle.LowPrice < _dailyLow)
			_dailyLow = candle.LowPrice;
	}

	private bool CanOpenNewTrade()
	{
		if (TradeVolume <= 0m)
			return false;

		if (Position != 0m)
			return false;

		return !HasActiveOrders();
	}

	private bool ShouldEnterLong(decimal closePrice, decimal longRange, decimal requiredRange, int hour)
	{
		if (hour >= BuyCutoffHour)
			return false;

		if (closePrice <= _dailyOpen)
			return false;

		if (longRange <= 0m)
			return false;

		if (requiredRange > 0m)
			return longRange > requiredRange;

		return true;
	}

	private bool ShouldEnterShort(decimal closePrice, decimal shortRange, decimal requiredRange, int hour)
	{
		if (hour >= SellCutoffHour)
			return false;

		if (closePrice >= _dailyOpen)
			return false;

		if (shortRange <= 0m)
			return false;

		if (requiredRange > 0m)
			return shortRange > requiredRange;

		return true;
	}

	private decimal GetRequiredRange()
	{
		if (MinimumRangePips <= 0m)
			return 0m;

		var step = Security?.PriceStep ?? Security?.Step ?? 0m;
		if (step <= 0m)
			return 0m;

		return step * MinimumRangePips;
	}

	private bool HasActiveOrders()
	{
		return Orders.Any(o => o.State.IsActive());
	}
}

File: API/4201-4300/4210_Huge_Income/PY/huge_income_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Strategies import Strategy
from StockSharp.Algo.Indicators import ExponentialMovingAverage, AverageTrueRange

class huge_income_strategy(Strategy):
    def __init__(self):
        super(huge_income_strategy, self).__init__()
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(8))) \
            .SetDisplay("Candle Type", "Timeframe.", "General")
        self._ema_length = self.Param("EmaLength", 20) \
            .SetDisplay("EMA Length", "Trend filter.", "Indicators")
        self._atr_length = self.Param("AtrLength", 14) \
            .SetDisplay("ATR Length", "ATR period.", "Indicators")
        self._prev_close = 0.0
        self._entry_price = 0.0

    @property
    def CandleType(self):
        return self._candle_type.Value
    @property
    def EmaLength(self):
        return self._ema_length.Value
    @property
    def AtrLength(self):
        return self._atr_length.Value

    def OnStarted2(self, time):
        super(huge_income_strategy, self).OnStarted2(time)
        self._prev_close = 0.0
        self._entry_price = 0.0
        self._ema = ExponentialMovingAverage()
        self._ema.Length = self.EmaLength
        self._atr = AverageTrueRange()
        self._atr.Length = self.AtrLength
        subscription = self.SubscribeCandles(self.CandleType)
        subscription.Bind(self._ema, self._atr, self.ProcessCandle).Start()

    def ProcessCandle(self, candle, ema_val, atr_val):
        if candle.State != CandleStates.Finished:
            return
        ev = float(ema_val)
        av = float(atr_val)
        close = float(candle.ClosePrice)
        if av <= 0 or self._prev_close == 0:
            self._prev_close = close
            return
        if self.Position > 0:
            if close < ev or close <= self._entry_price - av * 2.0:
                self.SellMarket()
                self._entry_price = 0.0
        elif self.Position < 0:
            if close > ev or close >= self._entry_price + av * 2.0:
                self.BuyMarket()
                self._entry_price = 0.0
        if not self.IsFormedAndOnlineAndAllowTrading():
            self._prev_close = close
            return
        if self.Position == 0:
            if close > ev and self._prev_close <= ev:
                self._entry_price = close
                self.BuyMarket()
            elif close < ev and self._prev_close >= ev:
                self._entry_price = close
                self.SellMarket()
        self._prev_close = close

    def OnReseted(self):
        super(huge_income_strategy, self).OnReseted()
        self._prev_close = 0.0
        self._entry_price = 0.0

    def CreateClone(self):
        return huge_income_strategy()
